lib/riemann/tools/http_check.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+require 'net/http'
+require 'resolv'
+
+require 'riemann/tools'
+
+# Test for HTTP requests
+module Riemann
+  module Tools
+    class HttpCheck
+      include Riemann::Tools
+      require 'socket'
+
+      opt :uri, 'URI to fetch', short: :none, type: :strings
+      opt :response, 'Expected response codes', short: :none, type: :strings, default: [
+        '200', # OK
+        '301', # Moved Permanently
+        '401', # Unauthorized
+      ]
+
+      def tick
+        opts[:uri].each do |uri|
+          test_uri(uri)
+        end
+      end
+
+      def test_uri(uri)
+        uri = URI(uri)
+
+        req = ::Net::HTTP::Get.new(uri)
+        req.basic_auth(uri.user, uri.password)
+
+        result = {}
+
+        Resolv::DNS.new.each_address(uri.host) do |address|
+          response = ::Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https', verify_mode: OpenSSL::SSL::VERIFY_NONE) do |http|
+            http.request(req)
+          end
+
+          result[address] = response
+        end
```

Comment:
Not sure what is better here:

Let's say we want to monitor 3 services, each has 2 IP addresses. One of these IP addresses is down (192.168.0.4) and  some website return different content to each get request (random.example.com).

### Aggregating the results

Sample events:

```
get https://example.com ok
get https://example.net critical
get https://random.example.com critical
```

Pro:
* We can compare each result and report an error when they are not consistent (random.example.com)

Con:
* We only have a all or nothing info: works or does not work. For example.net, we can only see what happen in the description of the event (not visible above) and can't graph this conveniently.

### Not aggregating the results

Sample events:

```
get https://example.com at 192.168.0.1:443 ok
get https://example.com at 192.168.0.2:443 ok
get https://example.net at 192.168.0.3:443 ok
get https://example.net at 192.168.0.4:443 critical
get https://random.example.com at 192.168.0.5:443 ok
get https://random.example.com at 192.168.0.6:443 ok
```

Pro:
* We have a more fine-grained information about failing endpoints (192.168.0.4:443)

Con:
* We cannot detect inconsistent results

### Both?

Maybe we can do both and have at least 2 events per endpoint? But that does not look so good when a service has only one IP address.


I don't know what to think about this :smile: . What do you think?

Reply:
I added a commit to have both:
* One event / IP address with the actual success of this request;
* One consistency event for the overall consistency overall IP addresses.